From 2b0e9c3082739c9fb2a7cbba8b8844a882944b81 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@freescale.com>
Date: Tue, 27 Jan 2015 15:01:57 +0800
Subject: [PATCH 0087/5242] MLK-10101-3 usb: chipidea: otg: enable BSV irq
 when OTG B-device in host mode

commit  06f9c388cfe0afd89fc051d16b4ee7a03caf831c from
https://source.codeaurora.org/external/imx/linux-imx.git

When B-device in host mode, if Vbus is off by A-device or A-device is removed,
B-device should update charger status correctly. This patch enables BSV irq
for B-device in all states, so the charger connection and removal can be early
handled by BSV change irq.

Signed-off-by: Li Jun <jun.li@freescale.com>
(cherry picked from commit eb7c192ce6ce077471dfdcdc4eb6cc9da4a7c01a)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/udc.c |   15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/chipidea/udc.c b/drivers/usb/chipidea/udc.c
index 1120969..aaa1a93 100644
--- a/drivers/usb/chipidea/udc.c
+++ b/drivers/usb/chipidea/udc.c
@@ -2026,8 +2026,12 @@ void ci_hdrc_gadget_connect(struct usb_gadget *gadget, int is_active)
 
 static int udc_id_switch_for_device(struct ci_hdrc *ci)
 {
-	if (ci->is_otg)
-		/* Clear and enable BSV irq */
+	if (!ci->is_otg)
+		return 0;
+
+	/* Clear and enable BSV irq for peripheral or OTG B-device */
+	if (!ci_otg_is_fsm_mode(ci) ||
+			ci->fsm.otg->state <= OTG_STATE_B_HOST)
 		hw_write_otgsc(ci, OTGSC_BSVIS | OTGSC_BSVIE,
 					OTGSC_BSVIS | OTGSC_BSVIE);
 
@@ -2036,11 +2040,14 @@ static int udc_id_switch_for_device(struct ci_hdrc *ci)
 
 static void udc_id_switch_for_host(struct ci_hdrc *ci)
 {
+	if (!ci->is_otg)
+		return;
 	/*
-	 * host doesn't care B_SESSION_VALID event
+	 * Host or OTG A-device doesn't care B_SESSION_VALID event
 	 * so clear and disbale BSV irq
 	 */
-	if (ci->is_otg)
+	if (!ci_otg_is_fsm_mode(ci) ||
+			ci->fsm.otg->state > OTG_STATE_B_HOST)
 		hw_write_otgsc(ci, OTGSC_BSVIE | OTGSC_BSVIS, OTGSC_BSVIS);
 
 	ci->vbus_active = 0;
-- 
1.7.9.5

